[warmup] Add Support for Warmup 4iE Device [WIP]

jamesmelville opened the pull request:
This binding integrates the Warmup 4iE Thermostat https://www.warmup.co.uk/thermostats/smart/4ie-underfloor-heating, via the API at https://my.warmup.com/.

I logged an issue for this at #7996 .

Per the readme, it currently supports a Bridge to the API, holding the API credentials, and Rooms, which report current temperature and setpoint, and allow setting of an override of the setpoint for a specified duration.

I have tested this successfully on my own installation.

There is a suggested todo list in the readme, but I don't regard any of the suggested functionality as essential and hope this could be merged as-is.

fwolter commented on bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,102 @@
+			<channel id="runMode" typeId="runMode"/>
+		</channels>
+
+		<representation-property>serialNumber</representation-property>
```
You used the Id in the discovery service. The serial number might be more reasonable.

jamesmelville replied:
Thanks - I think this is all aligned now

fwolter commented on bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,102 @@
+		<state min="5" max="30" step="0.5" readOnly="false" pattern="%.1f %unit%"/>
+	</channel-type>
+
+	<channel-type id="overrideRemaining">
```
Is there a reason why you didn't document these Channels?

jamesmelville replied:
No reason, I've added them to the updated readme

jamesmelville commented on bundles/org.openhab.binding.warmup/README.md:
```diff
@@ -0,0 +1,68 @@
+## Supported Things
+
+* Bridge - Credentials to my.warmup.com which allow the rest of the system to function
+* Room - An individual Warmup 4iE device which is a WiFi connected device which controls a heating circuit. The device is optimised for controlling underfloor heating (electric or hydronic), although it can also control central heating circuits. The device reports the temperature from one of two thermostats, either a floor temperature probe or the air temperature at the device. The separate tempe ...
```
done

jamesmelville commented on bundles/org.openhab.binding.warmup/README.md:
````diff
@@ -0,0 +1,68 @@
+}
+```
+
+## Todo
````
removed - is there anywhere you'd recommend to keep this? They're more ideas for if other people want to put effort into extending the binding than a todo list for me. I've focused on the currently implemented features as a complete usable MVP rather than a full client.

fwolter replied:
You could open a GitHub issue in this repo.

jamesmelville commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupApi.java:
```diff
@@ -0,0 +1,183 @@
+                authToken = null;
+            } else {
+            }
+        } catch (Exception e) {
```
I've reworked the error handling throughout which I hope resolves all the concerns you've highlighted, any feedback will be much appreciated!

jamesmelville commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupConfigurationDTO.java:
```diff
@@ -0,0 +1,29 @@
+    public String password;
+
+    @Override
+    public int compareTo(MyWarmupConfigurationDTO o) {
```
I'd initially put this in so that the existing session token was not discarded if the username/pw in the configuration were unchanged when the bridge was reconfigured. I've now taken it out because it's probably unnecessary complexity and I've not seen the API get unhappy with frequent logins.

jamesmelville commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java:
```diff
@@ -0,0 +1,84 @@
+                thingDiscovered(DiscoveryResultBuilder.create(new ThingUID(THING_TYPE_ROOM, bridgeUID, deviceSN))
+                        .withBridge(bridgeUID).withProperties(roomProperties)
+                        .withLabel(location.getName() + " - " + room.getName())
+                        .withRepresentationProperty(room.getId().toString()).build());
```
Resolved - thank you for correcting my misunderstanding of this method.

jamesmelville commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java:
```diff
@@ -0,0 +1,84 @@
+    public void onRefresh(final QueryResponseDTO domain) {
+        for (LocationDTO location : domain.getData().getUser().getLocations()) {
+            for (RoomDTO room : location.getRooms()) {
+                String deviceSN = room.getThermostat4ies().get(0).getDeviceSN();
```
Resolved

jamesmelville commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java:
```diff
@@ -0,0 +1,128 @@
+        updateStatus(ThingStatus.UNKNOWN);
+        api.setConfiguration(getConfigAs(MyWarmupConfigurationDTO.class));
+        discoveryService = new WarmupDiscoveryService(this);
+        BigDecimal refreshInterval = (BigDecimal) getConfig().get("refreshInterval");
```
Configuration DTOs are now used throughout

jamesmelville commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java:
```diff
@@ -0,0 +1,128 @@
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
```
I removed this - the status is now set once a first attempt to call API is completed

jamesmelville commented on bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,102 @@
+				<label>Username</label>
+				<description>Username for my.warmup.com</description>
+			</parameter>
+			<parameter name="password" type="text" required="true">
```
added

jamesmelville commented on bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Warmup Binding</name>
+	<description>This is the binding for Warmup.</description>
```
I extended the description, I hope it's now suitable

jamesmelville commented:
Hi Fabian @fwolter, thanks for the review. I've updated everything you requested, fixed the checkstyle warnings, and my build is now passing. I've only had one owner of a device provide feedback https://community.openhab.org/t/warmup-thermostat-41e/37394/5, but I'll see if I can get any more.

fwolter commented on bundles/org.openhab.binding.warmup/README.md:
```diff
@@ -0,0 +1,86 @@
+| currentTemperature | Number:Temperature | Currently reported temperature | true |
+| targetTemperature | Number:Temperature | Target temperature | false |
+| overrideRemaining | Number:Time | Duration remaining of the configured override| true |
+| runMode | String | Current operating mode of the thermostat | true |
```
Can you mention these modes?

jamesmelville replied:
I've added further explanation (based on my limited knowledge!).

fwolter commented on bundles/org.openhab.binding.warmup/README.md:
```diff
@@ -0,0 +1,86 @@
+| targetTemperature | Number:Temperature | Target temperature | false |
+| overrideRemaining | Number:Time | Duration remaining of the configured override| true |
+| runMode | String | Current operating mode of the thermostat | true |
+| frostProtectionMode | Switch | Toggles between the "Frost Protection" run mode and the previously configured mode | false |
```
Do you mean the mode in `runMode` with "previously configured mode"?

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupApi.java:
```diff
@@ -0,0 +1,202 @@
+        } catch (MyWarmupApiException e) {
+            failCount++;
+            if (failCount > 2) {
+                logger.error("Multiple authentication failures: {}", e.getMessage());
```
Bindings should only log to error if something severe happened, like the detection of a bug in your code. This can be warn.

Also, it's good practice to log exceptions where they are catched. This prevents multiple logging of the same exception in different locations.

jamesmelville replied:
Thanks, updated

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java:
```diff
@@ -0,0 +1,126 @@
+    public void initialize() {
+        MyWarmupConfigurationDTO config = getConfigAs(MyWarmupConfigurationDTO.class);
+        api.setConfiguration(config);
+        refreshJob = scheduler.scheduleWithFixedDelay(this::refreshFromServer, 0, config.refreshIntervalSec,
```
Better validate the value of `config.refreshIntervalSec` as it can be 0 if the user doesn't configure it. Then, an IllegalArgumentException would be raised.

jamesmelville replied:
Implemented guard against value being below 10 specified in thing.xml https://github.com/openhab/openhab-addons/pull/7998/commits/7bc9ed18881442a51c73f95fe123773f6d04e482#diff-028f0585e493ccac6de2540a9f878ba5

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java:
```diff
@@ -0,0 +1,133 @@
+
+    public RoomHandler(Thing thing) {
+        super(thing);
+        config = getConfigAs(RoomConfigurationDTO.class);
```
The configuration should be read in `initialize()`, since the handler object is re-used when the configuration is updated. Then, `dispose()` followed by `initialize()` will be invoked, without calling the constructor.

jamesmelville replied:
Thank you, this is resolved, although I hope the @SuppressWarnings("null") I added is ok? https://github.com/openhab/openhab-addons/pull/7998/commits/7bc9ed18881442a51c73f95fe123773f6d04e482#diff-782a12acc9c698d8cf1b524ad394758e

fwolter replied:
Sorry for the delay. Please do a null check instead of suppressing the warning. I'm going to approve the code after the fix. @Hilbrand you could do a second review if you want.

jamesmelville replied:
resolved in https://github.com/openhab/openhab-addons/pull/8562/commits/7be6f8d8ab878b0c37f5040555ce9e4f9707188c

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupThingHandler.java:
```diff
@@ -0,0 +1,112 @@
+     */
+    protected State parseDate(@Nullable String date, @Nullable String timezone) {
+        return date != null && timezone != null ? new DateTimeType(ZonedDateTime
+                .of(LocalDateTime.parse(date, DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")), ZoneId.of(timezone)))
```
You could make `DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")` a static final constant to save resources.

jamesmelville replied:
Code removed as it was unused

Hilbrand commented on bundles/org.openhab.binding.warmup/README.md:
```diff
@@ -0,0 +1,104 @@
+
+Any Warmup 4iE device(s) must be registered at https://my.warmup.com/ prior to usage.
+
+This API is not known to be documented publicly. The binding api implementation has been derived from the implementations at https://github.com/alyc100/SmartThingsPublic/blob/master/devicetypes/alyc100/warmup-4ie.src/warmup-4ie.groovy and https://github.com/alex-0103/warmup4IE/blob/master/warmup4ie/warmup4ie.py, and enhanced by inspecting the GraphQL endpoint.
```
Can you put each sentence on it's own line, that makes viewing (future) diffs more easily.
```suggestion
This API is not known to be documented publicly.
The binding api implementation has been derived from the implementations at https://github.com/alyc100/SmartThingsPublic/blob/master/devicetypes/alyc100/warmup-4ie.src/warmup-4ie.groovy and https://github.com/alex-0103/warmup4IE/blob/master/warmup4ie/warmup4ie.py, and enhanced by inspecting the GraphQL endpoint.
```

Hilbrand commented on bundles/org.openhab.binding.warmup/README.md:
```diff
@@ -0,0 +1,104 @@
+
+### My Warmup Account
+
+| name  | type   | description                  | required | default |
+| --------- | -------- | ------------------------------ | ----------- | -------- |
+| username | String | Username for my.warmup.com | true | |
+| password | String | Password for my.warmup.com | true | |
+| refreshIntervalSec | Integer | Interval in seconds between automatic refreshes | true | 300 |
```
Can you format the tables. You can use http://markdowntable.com/

Hilbrand commented on bundles/org.openhab.binding.warmup/pom.xml:
```diff
@@ -0,0 +1,17 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>2.5.8-SNAPSHOT</version>
```
Can you rebase to 2.5.9. With `git pull --rebase` and than `git push --force-with-lease` to get the changes back to GitHub
```suggestion
    <version>2.5.9-SNAPSHOT</version>
```

jamesmelville replied:
Migrated to v3.0.0

Hilbrand commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupHandlerFactory.java:
```diff
@@ -0,0 +1,92 @@
+
+        if (THING_TYPE_BRIDGE.equals(thingTypeUID)) {
+            MyWarmupAccountHandler bridge = new MyWarmupAccountHandler((Bridge) thing, httpClient);
+            registerDiscoveryService(bridge);
```
Can you implement discovery registration as `ThingHandlerService` See this preview documentation about how that looks like: https://deploy-preview-1262--openhab-docs-preview.netlify.app/docs/developer/bindings/#discovery-that-is-bound-to-a-thing

jamesmelville replied:
Resolved in https://github.com/openhab/openhab-addons/pull/8562/commits/4c41f60f5463036c3e0300170cf496166db005c7

Hilbrand commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java:
```diff
@@ -0,0 +1,99 @@
+                }
+            }
+
+            previousThings.removeAll(discoveredThings);
+            for (ThingUID missingThing : previousThings) {
+                thingRemoved(missingThing);
+            }
+            previousThings = discoveredThings;
```
In general older results are removed by adding the follow line to the `stopScan()` method:
```
removeOlderResults(getTimestampOfLastScan());
```

Hilbrand commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    private void notifyListeners(@Nullable QueryResponseDTO domain) {
+        if (discoveryService != null && queryResponse != null) {
+            discoveryService.refresh(queryResponse);
+        }
```
A discovery should only be triggered from the discovery service itself. So this part might not be a good idea. What is the reason to call it here?

Hilbrand commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupThingHandler.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Override
+    public void initialize() {
```
You should probably set the status of the thing to `UNKNOWN` here, or is the status handled elsewhere?

Hilbrand reviewed (changes requested):
Thank you for your contribution. Just a few comments, mostly related to discovery.

Make sure you run `mvn spotless:apply` on you binding before pushing the code to GitHub.

If you're able to process these review comments in the next days we can add this new binding to 2.5.9, otherwise you'll have to migrate it to openHAB 3.0.

jamesmelville commented:
Hi @Hilbrand @fwolter, thank you for your reviews. Unfortunately I've been on holiday so missed the deadline for 2.5.9. I've made all the changes you've requested along with updating the pull request for 3.0.0 here #8562.

fwolter commented:
Closing in favor of #8562
